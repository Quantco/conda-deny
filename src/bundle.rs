use scopeguard::defer;
use std::fs::{self, create_dir_all, File};
use std::io::{self, copy, BufReader};
use std::path::{Path, PathBuf};

use bzip2::read::BzDecoder;

use anyhow::{Context, Result};
use log::{debug, warn};
use reqwest::blocking::get;
use tar::Archive;
use zip::ZipArchive;
use zstd::Decoder;

type LicenseFileName = String;
type LicenseText = String;

pub fn get_license_contents_for_package_url(url: &str) -> Result<Vec<(String, String)>> {
    let file_name = Path::new(url)
        .file_name()
        .and_then(|name| name.to_str())
        .ok_or_else(|| anyhow::anyhow!("Failed to extract file name from URL"))?;

    let output_dir = Path::new(file_name)
        .file_stem()
        .and_then(|stem| stem.to_str())
        .ok_or_else(|| anyhow::anyhow!("Failed to get file stem as str"))?;

    download_file(url, file_name)?;
    defer! {
        let _ = fs::remove_file(file_name);
    }
    unpack_conda_file(file_name)?;
    defer! {
        let _ = fs::remove_dir_all(output_dir);
    }
    let license_strings = get_licenses_from_unpacked_conda_package(output_dir)?;

    std::fs::remove_file(file_name)
        .with_context(|| format!("Failed to delete file {}", file_name))?;
    std::fs::remove_dir_all(output_dir)
        .with_context(|| format!("Failed to remove directory {}", output_dir))?;

    Ok(license_strings)
}

fn find_all_licenses_directories(root: &Path) -> Result<Vec<PathBuf>> {
    let mut licenses_dirs = Vec::new();
    visit_dir(root, &mut licenses_dirs)?;
    Ok(licenses_dirs)
}

fn visit_dir(path: &Path, licenses_dirs: &mut Vec<PathBuf>) -> Result<()> {
    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let entry_path = entry.path();

        if entry_path.is_dir() {
            if entry_path.file_name().unwrap() == "licenses" {
                licenses_dirs.push(entry_path.clone());
            } else {
                visit_dir(&entry_path, licenses_dirs)?;
            }
        }
    }
    Ok(())
}

fn get_licenses_from_unpacked_conda_package(
    unpacked_conda_package_dir: &str,
) -> Result<Vec<(LicenseFileName, LicenseText)>> {
    let mut license_strings = Vec::new();

    let licenses_dirs = find_all_licenses_directories(Path::new(unpacked_conda_package_dir))?;

    if !licenses_dirs.is_empty() {
        for licenses_dir in licenses_dirs {
            get_license_texts_for_dir(&licenses_dir, &mut license_strings).with_context(|| {
            format!(
                "Failed to get license content from {}. Does the licenses directory exist within the package?",
                licenses_dir.display()
            )
        })?;
        }
        if license_strings.is_empty() {
            warn!(
                "Warning: No license files found in {}. Adding default license message.",
                unpacked_conda_package_dir
            );
            license_strings.push((
                "NO LICENSE FOUND".to_string(),
                "THE LICENSE OF THIS PACKAGE IS NOT PACKAGED!".to_string(),
            ));
        }
    } else {
        warn!(
            "Warning: No 'info/licenses' directory found in {}. Adding default license message.",
            unpacked_conda_package_dir
        );
        license_strings.push((
            "NO LICENSE FOUND".to_string(),
            "THE LICENSE OF THIS PACKAGE IS NOT PACKAGED!".to_string(),
        ));
    }

    license_strings.sort();
    license_strings.dedup();

    Ok(license_strings)
}

fn get_license_texts_for_dir(
    path: &Path,
    license_strings: &mut Vec<(LicenseFileName, LicenseText)>,
) -> Result<()> {
    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let entry_path = entry.path();

        if entry_path.is_dir() {
            get_license_texts_for_dir(&entry_path, license_strings)?;
        } else {
            let entry_file_name = entry.file_name().to_string_lossy().to_string();
            let content = fs::read_to_string(&entry_path)
                .with_context(|| format!("Failed to read {:?}", entry_path))?;
            license_strings.push((entry_file_name, content));
        }
    }
    Ok(())
}

fn download_file(url: &str, file_path: &str) -> Result<()> {
    let response = get(url).with_context(|| format!("Failed to download {}", file_path))?;

    if !response.status().is_success() {
        return Err(anyhow::anyhow!(
            "Failed to download file: {}",
            response.status()
        ));
    }

    let mut dest = File::create(file_path)
        .with_context(|| format!("File at {} could not be created", file_path))?;
    let content = response.bytes()?;

    copy(&mut content.as_ref(), &mut dest)?;

    debug!("File downloaded successfully to {}", file_path);
    Ok(())
}

fn unpack_conda_file(file_path: &str) -> Result<()> {
    let output_dir = Path::new(file_path)
        .file_stem()
        .map(PathBuf::from)
        .expect("Failed to get file stem");

    let file_extension = Path::new(file_path)
        .extension()
        .and_then(|ext| ext.to_str())
        .unwrap_or("");

    match file_extension {
        "conda" => unpack_conda_archive(file_path, &output_dir),
        "bz2" => unpack_tar_bz2_archive(file_path, &output_dir),
        other => Err(anyhow::anyhow!(format!(
            "Unsupported file extension: {}",
            other
        ))),
    }
}

fn unpack_conda_archive(file_path: &str, output_dir: &Path) -> Result<()> {
    let zip_file =
        File::open(file_path).with_context(|| format!("Failed to open {}", file_path))?;
    let mut zip = ZipArchive::new(BufReader::new(zip_file))
        .with_context(|| "Failed to create zip archive")?;

    for i in 0..zip.len() {
        let mut zip_file = zip.by_index(i)?;
        if zip_file.name().ends_with(".tar.zst") {
            let mut tar_zst_data = Vec::new();
            io::copy(&mut zip_file, &mut tar_zst_data)?;

            let mut decoder = Decoder::new(&tar_zst_data[..])?;
            let mut tar_data = Vec::new();
            io::copy(&mut decoder, &mut tar_data)?;

            let mut tar = Archive::new(&tar_data[..]);
            create_dir_all(output_dir).with_context(|| {
                format!(
                    "Failed to create directory {}",
                    output_dir.to_string_lossy()
                )
            })?;
            tar.unpack(output_dir)
                .with_context(|| format!("Failed to unpack {}", output_dir.to_string_lossy()))?;
            debug!("Successfully unpacked to {:?}", output_dir);
        }
    }
    Ok(())
}

fn unpack_tar_bz2_archive(file_path: &str, output_dir: &Path) -> Result<()> {
    let tar_bz2_file =
        File::open(file_path).with_context(|| format!("Failed to open {}", file_path))?;
    let bz2_decoder = BzDecoder::new(tar_bz2_file);
    let mut tar = Archive::new(bz2_decoder);

    create_dir_all(output_dir).with_context(|| {
        format!(
            "Failed to create directory {}",
            output_dir.to_string_lossy()
        )
    })?;
    tar.unpack(output_dir)
        .with_context(|| format!("Failed to unpack {}", output_dir.to_string_lossy()))?;
    debug!("Successfully unpacked .tar.bz2 to {:?}", output_dir);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_all_licenses_directories() {
        let root = Path::new("tests/test_bundle_data/polarify-0.2.0-pyhd8ed1ab_0.conda");
        let result = find_all_licenses_directories(root).unwrap();

        assert_eq!(result.len(), 2);
        assert!(result.contains(&PathBuf::from("tests/test_bundle_data/polarify-0.2.0-pyhd8ed1ab_0.conda/pkg-polarify-0.2.0-pyhd8ed1ab_0/site-packages/polarify-0.2.0.dist-info/licenses")));
        assert!(result.contains(&PathBuf::from("tests/test_bundle_data/polarify-0.2.0-pyhd8ed1ab_0.conda/pkg-polarify-0.2.0-pyhd8ed1ab_0/info/licenses")));

        let root = Path::new("tests/test_bundle_data/_libgcc_mutex-0.1-free");
        let result = find_all_licenses_directories(root).unwrap();

        assert_eq!(result.len(), 0);
    }

    #[test]
    fn test_get_licenses_from_unpacked_conda_package_with_license_files() {
        let unpacked_conda_dir =
            Path::new("tests/test_bundle_data/polarify-0.2.0-pyhd8ed1ab_0.conda");

        let result =
            get_licenses_from_unpacked_conda_package(unpacked_conda_dir.to_str().unwrap()).unwrap();

        assert_eq!(result.len(), 1);
        assert!(result.contains(&(
            String::from("LICENSE"),
            String::from("This is the license.")
        )));
    }

    #[test]
    fn test_get_licenses_from_unpacked_conda_package_without_licenses_directory() {
        let unpacked_conda_dir = Path::new("tests/test_bundle_data/_libgcc_mutex-0.1-free");

        let result =
            get_licenses_from_unpacked_conda_package(unpacked_conda_dir.to_str().unwrap()).unwrap();

        assert_eq!(result.len(), 1);
        assert!(result.contains(&(
            "NO LICENSE FOUND".to_string(),
            "THE LICENSE OF THIS PACKAGE IS NOT PACKAGED!".to_string()
        )));
    }

    #[test]
    fn test_get_licenses_from_unpacked_conda_package_empty_liceses_dir() {
        let unpacked_conda_dir = Path::new("tests/test_bundle_data/empty_licenses_dir");

        let result =
            get_licenses_from_unpacked_conda_package(unpacked_conda_dir.to_str().unwrap()).unwrap();

        assert_eq!(result.len(), 1);
        assert!(result.contains(&(
            "NO LICENSE FOUND".to_string(),
            "THE LICENSE OF THIS PACKAGE IS NOT PACKAGED!".to_string()
        )));
    }
}
