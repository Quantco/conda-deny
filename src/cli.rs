use clap::ArgAction;
use clap_verbosity_flag::{ErrorLevel, Verbosity};

use clap::Parser;

#[derive(Parser, Debug)]
#[command(name = "conda-deny", about = "Check and list licenses of pixi and conda environments", version = env!("CARGO_PKG_VERSION"))]
pub struct Cli {
    #[command(flatten)]
    pub verbose: Verbosity<ErrorLevel>,

    #[command(subcommand)]
    pub command: CondaDenyCliConfig,

    /// Path to the conda-deny config file
    #[arg(short, long, global = true)]
    pub config: Option<String>,
}

#[derive(clap::Subcommand, Debug)]
pub enum CondaDenyCliConfig {
    Check {
        /// Path to the pixi lockfile(s)
        #[arg(short, long)]
        lockfile: Option<Vec<String>>,

        /// Path to the conda prefix(es)
        #[arg(long, global = true)]
        prefix: Option<Vec<String>>,

        /// Platform(s) to check
        #[arg(short, long)]
        platform: Option<Vec<String>>,

        /// Environment(s) to check
        #[arg(short, long)]
        environment: Option<Vec<String>>,

        #[arg(short, long, action = ArgAction::SetTrue)]
        include_safe: bool,

        #[arg(short, long)]
        osi: Option<bool>,

        /// Ignore when encountering pypi packages instead of failing.
        #[arg(long)]
        ignore_pypi: Option<bool>,
    },
    List {
        /// Path to the pixi lockfile(s)
        #[arg(short, long)]
        lockfile: Option<Vec<String>>,

        /// Path to the conda prefix(es)
        #[arg(long, global = true)]
        prefix: Option<Vec<String>>,

        /// Platform(s) to list
        #[arg(short, long)]
        platform: Option<Vec<String>>,

        /// Environment(s) to list
        #[arg(short, long)]
        environment: Option<Vec<String>>,

        /// Ignore when encountering pypi packages instead of failing.
        #[arg(long)]
        ignore_pypi: Option<bool>,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cli_check_include_safe() {
        let cli = Cli::try_parse_from(vec!["conda-deny", "check", "--include-safe"]).unwrap();
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(include_safe);
            }
            _ => panic!("Expected check subcommand with --include-safe"),
        }
    }

    #[test]
    fn test_cli_with_config() {
        let cli =
            Cli::try_parse_from(vec!["conda-deny", "list", "--config", "custom.toml"]).unwrap();
        assert_eq!(cli.config.as_deref(), Some("custom.toml"));
    }

    #[test]
    fn test_cli_with_config_new_order() {
        let cli =
            Cli::try_parse_from(vec!["conda-deny", "check", "--config", "custom.toml"]).unwrap();
        assert_eq!(cli.config.as_deref(), Some("custom.toml"));
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(!include_safe);
            }
            _ => panic!("Expected check subcommand with --config"),
        }
    }

    #[test]
    fn test_cli_with_check_arguments() {
        let cli = Cli::try_parse_from(vec!["conda-deny", "check", "--include-safe"]).unwrap();
        match cli.command {
            CondaDenyCliConfig::Check { include_safe, .. } => {
                assert!(include_safe);
            }
            _ => panic!("Expected check subcommand with --include-safe"),
        }
    }
}
