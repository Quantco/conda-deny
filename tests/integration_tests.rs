#![allow(clippy::too_many_arguments)]
use assert_cmd::prelude::*;
use conda_deny::bundle::bundle;
use conda_deny::cli::CondaDenyCliConfig;
use conda_deny::{
    check::check, get_config_options, list::list, CondaDenyCheckConfig, CondaDenyConfig,
    CondaDenyListConfig,
};
use conda_deny::{CondaDenyBundleConfig, OutputFormat};
use rattler_conda_types::Platform;
use rstest::{fixture, rstest};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::NamedTempFile;
use walkdir::WalkDir;

#[fixture]
fn list_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<String>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(Some(OutputFormat::Default))] output: Option<OutputFormat>,
) -> CondaDenyListConfig {
    let cli = CondaDenyCliConfig::List {
        lockfile,
        prefix,
        platform,
        environment,
        ignore_pypi,
        output,
    };

    let config = get_config_options(config, cli).unwrap();

    match config {
        CondaDenyConfig::List(list_config) => list_config,
        _ => panic!(),
    }
}

#[fixture]
fn bundle_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<String>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(None)] directory: Option<PathBuf>,
) -> CondaDenyBundleConfig {
    let cli = CondaDenyCliConfig::Bundle {
        lockfile,
        prefix,
        platform,
        environment,
        ignore_pypi,
        directory,
    };

    let config = get_config_options(config, cli).unwrap();

    match config {
        CondaDenyConfig::Bundle(bundle_config) => bundle_config,
        _ => panic!(),
    }
}

#[fixture]
fn check_config(
    #[default(None)] config: Option<PathBuf>,
    #[default(None)] lockfile: Option<Vec<String>>,
    #[default(None)] prefix: Option<Vec<PathBuf>>,
    #[default(None)] platform: Option<Vec<Platform>>,
    #[default(None)] environment: Option<Vec<String>>,
    #[default(None)] osi: Option<bool>,
    #[default(None)] ignore_pypi: Option<bool>,
    #[default(Some(OutputFormat::Default))] output: Option<OutputFormat>,
) -> CondaDenyCheckConfig {
    let cli = CondaDenyCliConfig::Check {
        lockfile,
        prefix,
        platform,
        environment,
        osi,
        ignore_pypi,
        output,
    };

    let config = get_config_options(config, cli);
    let config = config.unwrap();

    match config {
        CondaDenyConfig::Check(check_config) => check_config,
        _ => panic!(),
    }
}

#[fixture]
fn out() -> Vec<u8> {
    Vec::new()
}

#[rstest]
#[case("check", "test_default_use_case")]
#[case("list", "test_default_use_case")]
#[case("check", "test_default_use_case_pyproject")]
#[case("list", "test_default_use_case_pyproject")]
fn test_default_use_case(#[case] subcommand: &str, #[case] test_name: &str) {
    use core::str;

    use log::debug;

    let path_string = format!("tests/{test_name}");
    let test_dir = Path::new(path_string.as_str());

    let output = Command::cargo_bin("conda-deny")
        .unwrap()
        .arg(subcommand)
        .current_dir(test_dir)
        .env("CLICOLOR_FORCE", "1")
        .output()
        .expect("Failed to execute command");

    let stdout = str::from_utf8(&output.stdout).unwrap();
    debug!("Output: {stdout}");
    if subcommand == "check" {
        assert!(stdout.contains("There were \u{1b}[32m247\u{1b}[0m safe licenses and \u{1b}[31m301\u{1b}[0m unsafe licenses."), "{stdout}");
        output.assert().failure();
    } else {
        assert!(stdout.contains("\u{1b}[34mzstandard\u{1b}[0m \u{1b}[36m0.22.0\u{1b}[0m-\u{1b}[3;96mpy312h721a963_1\u{1b}[0m (\u{1b}[95mosx-arm64\u{1b}[0m): \u{1b}[33mBSD-3-Clause"));
        assert!(stdout.contains("\u{1b}[34mzlib\u{1b}[0m \u{1b}[36m1.3.1\u{1b}[0m-\u{1b}[3;96mh4ab18f5_1\u{1b}[0m (\u{1b}[95mlinux-64\u{1b}[0m): \u{1b}[33mZlib"));
        assert!(stdout.contains("\u{1b}[34mxz\u{1b}[0m \u{1b}[36m5.2.6\u{1b}[0m-\u{1b}[3;96mh166bdaf_0\u{1b}[0m (\u{1b}[95mlinux-64\u{1b}[0m): \u{1b}[33mLGPL-2.1 and GPL-2.0"));
        output.assert().success();
    }
}

#[rstest]
#[case("check")]
#[case("list")]
fn test_lockfile_pattern(#[case] subcommand: &str) {
    let test_dir = Path::new("tests/test_lockfile_pattern");
    let config = PathBuf::from(test_dir).join("pixi.toml");

    let mut out = out();

    let check_config = check_config(
        Some(config.clone()),
        None,
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let list_config = list_config(Some(config), None, None, None, None, None, None);

    if subcommand == "check" {
        let result = check(check_config, &mut out);
        let output = String::from_utf8(strip_ansi_escapes::strip(out)).unwrap();

        assert!(output.contains("There were 1 safe licenses and 21 unsafe licenses."));
        assert!(result.is_err());
    } else if subcommand == "list" {
        let result = list(list_config, &mut out);
        let output = String::from_utf8(strip_ansi_escapes::strip(out)).unwrap();

        // only in subdir lockfile
        assert!(output.contains("k9s 0.50.4-h643be8f_0 (linux-64): Apache-2.0"));
        // only in subdir/another_subdir lockfile
        assert!(output.contains("vhs 0.7.2-ha770c72_0 (linux-64): MIT"));
        assert!(result.is_ok())
    } else {
        panic!("Invalid subcommand");
    }
}

#[test]
fn test_remote_allowlist_check() {
    colored::control::set_override(false);
    
    // Create a temporary file for the license_allowlist.toml
    let mut temp_config_file = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-allowlist = "https://raw.githubusercontent.com/Quantco/conda-deny/refs/heads/main/tests/default_license_allowlist.toml""#;

    temp_config_file
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let temp_config_file_path = temp_config_file.path().to_path_buf();

    let mut out = out();
    let check_config = check_config(
        Some(temp_config_file_path),
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[test]
fn test_multiple_allowlists_check() {
    colored::control::set_override(false);
    // Create a temporary file for the license_allowlist.toml
    let mut temp_license_allowlist = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
                                safe-licenses = ["BSD-3-Clause"]"#;
    temp_license_allowlist
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let temp_allowlist_path = temp_license_allowlist.path().to_path_buf();

    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = "[tool.conda-deny]
        license-allowlist = [
        \"https://raw.githubusercontent.com/Quantco/conda-deny/refs/heads/main/tests/default_license_allowlist.toml\",
        \"".to_string() + temp_allowlist_path.to_str().unwrap() + "\"]";

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(
        temp_path,
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[test]
fn test_platform_env_restrictions_check() {
    colored::control::set_override(false);

    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-allowlist = "tests/default_license_allowlist.toml"
lockfile = "tests/default_pixi.lock"
platform = "linux-64"
environment = "lint""#;

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(temp_path, None, None, None, None, None, None, None);

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[test]
fn test_safe_licenses_in_config_check() {
    colored::control::set_override(false);

    // Create a temporary file for pixi.toml
    let mut temp_pixi_toml = NamedTempFile::new().unwrap();
    let file_content = r#"[tool.conda-deny]
license-allowlist = "tests/default_license_allowlist.toml"
safe-licenses = ["BSD-3-Clause"]"#;

    temp_pixi_toml
        .as_file_mut()
        .write_all(file_content.as_bytes())
        .unwrap();

    let mut out = out();
    // Inject the temporary file's path into check_config
    let temp_path = Some(temp_pixi_toml.path().to_path_buf());
    let check_config = check_config(
        temp_path,
        Some(vec!["tests/default_pixi.lock".into()]),
        None,
        None,
        None,
        None,
        None,
        None,
    );

    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[rstest]
fn test_osi_check(
    #[with(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        Some(vec!["tests/default_pixi.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        Some(true)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    colored::control::set_override(false);
    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[rstest]
fn test_prefix_list(
    #[with(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        None,
        // PREFIXES
        Some(vec!["tests/test_conda_prefixes/test-env".into()])
)]
    list_config: CondaDenyListConfig,
    mut out: Vec<u8>,
) {
    // When --prefix is specified, only the license information for the conda-meta directory in the specified prefix should be listed
    // License information from pixi.lock should not be listed
    colored::control::set_override(false);
    let result = list(list_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_ok(), "{:?}", result.unwrap_err());
    insta::assert_snapshot!(output);
}

#[rstest]
fn test_exception_check(
    #[with(Some(PathBuf::from(
        // CONFIG PATH
        "tests/test_exception_use_case/config_with_exception.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/default_pixi.lock".into()])
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    colored::control::set_override(false);
    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[rstest]
fn test_pypi_ignore_check(
    #[with(
        // CONFIG PATH
        Some(PathBuf::from("tests/test_pypi_ignore/pixi.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/test_pypi_ignore/lockfile_with_pypi_packages.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        None,
        // IGNORE PYPI
        Some(true)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    colored::control::set_override(false);
    let result = check(check_config, &mut out);
    let output = String::from_utf8(out).unwrap();

    assert!(result.is_err());
    insta::assert_snapshot!(output);
}

#[rstest]
fn test_pypi_ignore_error(
    #[with(
        // CONFIG PATH
        Some(PathBuf::from("tests/test_pypi_ignore/pixi.toml")),
        // LOCKFILE PATHS
        Some(vec!["tests/test_pypi_ignore/lockfile_with_pypi_packages.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // OSI FLAG
        None,
        // IGNORE PYPI
        Some(false)
    )]
    check_config: CondaDenyCheckConfig,
    mut out: Vec<u8>,
) {
    let result = check(check_config, &mut out);
    if let Err(e) = &result {
        println!("Actual error: {e}");
    }
    assert!(result.is_err());
    assert!(format!("{result:?}").contains("Pypi packages are not supported: beautifulsoup4"));
    assert_eq!(out, b"");
}

#[rstest]
fn test_bundle_prefix() {
    let mut out = out();
    let temp_dir = tempfile::tempdir().unwrap();
    let bundle_config = bundle_config(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        None,
        // PREFIXES
        Some(vec!["tests/test_conda_prefixes/test-env".into()]),
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // IGNORE PYPI
        None,
        // DIRECTORY
        Some(temp_dir.path().join(Path::new("test_bundle"))),
    );

    // Suppress progress bar
    std::env::set_var("NO_PROGRESS", "1");

    bundle(bundle_config.clone(), &mut out).unwrap();
    let bundle_dir = bundle_config.directory.unwrap();

    let mut entries = Vec::new();

    for e in WalkDir::new(&bundle_dir).into_iter().filter_map(|e| e.ok()) {
        let full_path = e.path();
        let rel_path = full_path.strip_prefix(&bundle_dir).unwrap().to_path_buf();
        entries.push(rel_path);
    }

    let mut entry_output = entries
        .iter()
        .map(|p| p.display().to_string())
        .collect::<Vec<_>>();

    entry_output.sort();
    let pretty_output = entry_output.join("\n");

    insta::assert_snapshot!(pretty_output);

    let output = String::from_utf8(out).unwrap();
    assert!(output.contains("License files written to:"));
}

#[rstest]
fn test_bundle_lockfile() {
    let mut out = out();
    let temp_dir = tempfile::tempdir().unwrap();
    let bundle_config = bundle_config(
        // CONFIG PATH
        None,
        // LOCKFILE PATHS
        Some(vec!["tests/test_default_use_case/pixi.lock".into()]),
        // PREFIXES
        None,
        // PLATFORM
        None,
        // ENVIRONMENT
        None,
        // IGNORE PYPI
        None,
        // DIRECTORY
        Some(temp_dir.path().join(Path::new("test_bundle"))),
    );

    // Suppress progress bar
    std::env::set_var("NO_PROGRESS", "1");

    bundle(bundle_config.clone(), &mut out).unwrap();
    let bundle_dir = bundle_config.directory.unwrap();

    let mut entries = Vec::new();

    for e in WalkDir::new(&bundle_dir).into_iter().filter_map(|e| e.ok()) {
        let full_path = e.path();
        let rel_path = full_path.strip_prefix(&bundle_dir).unwrap().to_path_buf();
        entries.push(rel_path);
    }

    let mut entry_output = entries
        .iter()
        .map(|p| p.display().to_string())
        .collect::<Vec<_>>();

    entry_output.sort();
    let pretty_output = entry_output.join("\n");

    insta::assert_snapshot!(pretty_output);

    let output = String::from_utf8(out).unwrap();
    assert!(output.contains("License files written to:"));
}
