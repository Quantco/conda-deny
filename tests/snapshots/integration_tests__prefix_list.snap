---
source: tests/integration_tests.rs
expression: output
---
asttokens 2.4.1-pyhd8ed1ab_0 (noarch): Apache-2.0 
bzip2 1.0.8-h99b78c6_7 (osx-arm64): bzip2-1.0.6 
ca-certificates 2024.8.30-hf0a4a13_0 (osx-arm64): ISC 
decorator 5.1.1-pyhd8ed1ab_0 (noarch): BSD-2-Clause 
exceptiongroup 1.2.2-pyhd8ed1ab_0 (noarch): MIT and PSF-2.0 
executing 2.1.0-pyhd8ed1ab_0 (noarch): MIT 
ipython 8.28.0-pyh707e725_0 (noarch): BSD-3-Clause 
jedi 0.19.1-pyhd8ed1ab_0 (noarch): MIT 
libblas 3.9.0-24_osxarm64_openblas (osx-arm64): BSD-3-Clause 
libcblas 3.9.0-24_osxarm64_openblas (osx-arm64): BSD-3-Clause 
libcxx 19.1.2-ha82da77_0 (osx-arm64): Apache-2.0 WITH LLVM-exception 
libexpat 2.6.3-hf9b8971_0 (osx-arm64): MIT 
libffi 3.4.2-h3422bc3_5 (osx-arm64): MIT 
libgfortran 5.0.0-13_2_0_hd922786_3 (osx-arm64): GPL-3.0-only WITH GCC-exception-3.1 
libgfortran5 13.2.0-hf226fd6_3 (osx-arm64): GPL-3.0-only WITH GCC-exception-3.1 
liblapack 3.9.0-24_osxarm64_openblas (osx-arm64): BSD-3-Clause 
libmpdec 4.0.0-h99b78c6_0 (osx-arm64): BSD-2-Clause 
libopenblas 0.3.27-openmp_h517c56d_1 (osx-arm64): BSD-3-Clause 
libsqlite 3.46.1-hc14010f_0 (osx-arm64): Unlicense 
libzlib 1.3.1-h8359307_2 (osx-arm64): Zlib 
llvm-openmp 19.1.2-hb52a8e5_0 (osx-arm64): Apache-2.0 WITH LLVM-exception 
matplotlib-inline 0.1.7-pyhd8ed1ab_0 (noarch): BSD-3-Clause 
multipledispatch 0.6.0-pyhd8ed1ab_1 (noarch): BSD-3-Clause 
ncurses 6.5-h7bae524_1 (osx-arm64): X11 AND BSD-3-Clause 
numpy 2.1.2-py313hab0c69d_0 (osx-arm64): BSD-3-Clause 
openssl 3.3.2-h8359307_0 (osx-arm64): Apache-2.0 
pandas 2.2.3-py313h47b39a6_1 (osx-arm64): BSD-3-Clause 
parso 0.8.4-pyhd8ed1ab_0 (noarch): MIT 
pexpect 4.9.0-pyhd8ed1ab_0 (noarch): ISC 
pickleshare 0.7.5-py_1003 (noarch): MIT 
pip 24.2-pyh145f28c_1 (noarch): MIT 
prompt-toolkit 3.0.48-pyha770c72_0 (noarch): BSD-3-Clause 
ptyprocess 0.7.0-pyhd3deb0d_0 (noarch): ISC 
pure_eval 0.2.3-pyhd8ed1ab_0 (noarch): MIT 
pygments 2.18.0-pyhd8ed1ab_0 (noarch): BSD-2-Clause 
python 3.13.0-h206b6c5_100_cp313 (osx-arm64): Python-2.0 
python-dateutil 2.9.0-pyhd8ed1ab_0 (noarch): Apache-2.0 
python-tzdata 2024.2-pyhd8ed1ab_0 (noarch): Apache-2.0 
python_abi 3.13-5_cp313 (osx-arm64): BSD-3-Clause 
pytz 2024.1-pyhd8ed1ab_0 (noarch): MIT 
readline 8.2-h92ec313_1 (osx-arm64): GPL-3.0-only 
six 1.16.0-pyh6c4a22f_0 (noarch): MIT 
stack_data 0.6.2-pyhd8ed1ab_0 (noarch): MIT 
tk 8.6.13-h5083fa2_1 (osx-arm64): TCL 
traitlets 5.14.3-pyhd8ed1ab_0 (noarch): BSD-3-Clause 
typing_extensions 4.12.2-pyha770c72_0 (noarch): PSF-2.0 
tzdata 2024b-hc8b5060_0 (noarch): LicenseRef-Public-Domain 
wcwidth 0.2.13-pyhd8ed1ab_0 (noarch): MIT 
xz 5.2.6-h57fd34a_0 (osx-arm64): LGPL-2.1 and GPL-2.0
